pkg/store/bucket.go

```diff
@@ -712,6 +775,131 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 	}, nil
 }
 
+// bucketBlockSet holds all blocks of an equal label set. It internally splits
+// them up by downsampling resolution and allows querying
+type bucketBlockSet struct {
+	labels      labels.Labels
+	mtx         sync.RWMutex
+	resolutions []int64          // available resolution, high to low
+	blocks      [][]*bucketBlock // ordered buckets for the existing resolutions
+}
+
+// newBucketBlockSet initializes a new set with the known downsampling windows hard-configured.
+// The set currently does not support arbtirary ranges.
+func newBucketBlockSet(lset labels.Labels) *bucketBlockSet {
+	return &bucketBlockSet{
+		labels:      lset,
+		resolutions: []int64{downsample.ResLevel2, downsample.ResLevel1, downsample.ResLevel0},
+		blocks:      make([][]*bucketBlock, 3),
+	}
+}
+
+func (s *bucketBlockSet) add(b *bucketBlock) error {
+	if !s.labels.Equals(labels.FromMap(b.meta.Thanos.Labels)) {
+		return errors.New("block's label set does not match set")
+	}
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	i := int64index(s.resolutions, b.meta.Thanos.Downsample.Resolution)
+	if i < 0 {
+		return errors.Errorf("unsupported downsampling resolution %d", b.meta.Thanos.Downsample.Resolution)
+	}
+	bs := append(s.blocks[i], b)
+	s.blocks[i] = bs
+
+	sort.Slice(bs, func(j, k int) bool {
+		if bs[j].meta.MinTime < bs[k].meta.MinTime {
+			return true
+		}
+		return false
+	})
+	return nil
+}
+
+func (s *bucketBlockSet) remove(id ulid.ULID) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	for i, bs := range s.blocks {
+		for j, b := range bs {
+			if b.meta.ULID != id {
+				continue
+			}
+			s.blocks[i] = append(bs[:j], bs[j+1:]...)
+			return
+		}
+	}
+}
+
+func int64index(s []int64, x int64) int {
+	for i, v := range s {
+		if v == x {
+			return i
+		}
+	}
+	return -1
+}
+
+// getFor returns a time-ordered list of blocks that cover date between mint and maxt.
+// Blocks with the lowest resolution possible but not lower than the given resolution are returned.
+func (s *bucketBlockSet) getFor(mint, maxt, minResolution int64) (bs []*bucketBlock) {
```

Comment:
how we got from `req.MaxResolutionWindow` to this `minResolution` name here? Also
```
for ; i < len(s.resolutions) && s.resolutions[i] > minResolution; i++ {}
```
Min resolution make sense here, but we are putting max Resolution in this parameter... Not sure what's going on here ^^

Reply:
That's essentially the issue about intuitively of naming we ran into before. `MaxResolutionWindow` and `minResolution` are in fact the same thing. If we set an upper bound on how big the window we aggregate over is that implies a minimum resolution we want to guarantee. If we made the window bigger, resolution would decrease.